pkg/api/errors/state.go

```diff
@@ -20,73 +20,136 @@ import (
 
 	"google.golang.org/grpc/codes"
 
-	kiterrors "github.com/dapr/kit/errors"
+	"github.com/dapr/components-contrib/metadata"
+	"github.com/dapr/kit/errors"
 )
 
-func StateStoreInvalidKeyName(storeName string, key string, msg string) error {
-	return kiterrors.NewBuilder(
-		codes.InvalidArgument,
-		http.StatusBadRequest,
-		msg,
-		"ERR_MALFORMED_REQUEST",
-	).WithErrorInfo(kiterrors.CodePrefixStateStore+kiterrors.CodeIllegalKey, nil).
-		WithResourceInfo("state", storeName, "", "").
-		WithFieldViolation(key, msg).
-		Build()
+type StateStoreError struct {
+	name             string
+	skipResourceInfo bool
+}
+
+func StateStore(name string) *StateStoreError {
+	return &StateStoreError{
+		name: name,
+	}
+}
+
+func (s *StateStoreError) NotFound(appID string) error {
+	msg := fmt.Sprintf("%s store %s is not found", metadata.StateStoreType, s.name)
+	s.skipResourceInfo = true
+	var meta map[string]string
+	if len(appID) > 0 {
+		meta = map[string]string{"appID": appID}
+	}
+	return s.build(
+		errors.NewBuilder(
+			codes.InvalidArgument,
+			http.StatusBadRequest,
+			msg,
+			"ERR_STATE_STORE_NOT_FOUND",
+		),
+		errors.CodeNotFound,
+		meta,
+	)
+}
+
+func (s *StateStoreError) NotConfigured(appID string) error {
```

Comment:
I wonder if it makes sense to still have a generic error. I can think of a few off the top of my head that apply to more than one of our APIs:

- NotConfigured
- Empty
- NotFound
- etc

Reply:
As discussed offline, this will be the forward moving implementation that each api will implement these generic errors for their api since the `trade off is totally fine given the benefit of things being typed.` We want to avoid having untyped string like params.